### PR TITLE
Draft a guide with our code quirks for new models

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR adds a few indications on code styling for new model additions. I put the one that I could think of quickly, but there are probably more!

### CODE REVIEW

File: docs/source/add_new_model.mdx [hunk 1]
```diff
@@ -95,6 +95,24 @@ different formats - the model to a *pytorch_model.bin* file and the configuratio
...
+### Code style
+
+When coding your new model, keep in mind that Transformers is an opinionated library and we have a few quirks of our
+own regarding how code should be writte :-)
+
+1. The forward pass of your model should be fully written in the modeling file while being fully independent of other
+   models in the library. If you want to reuse a block from another model, copy the code and paste it with a
+   `# Copied from` comment on top (see [here](https://github.com/huggingface/transformers/blob/v4.17.0/src/transformers/models/roberta/modeling_roberta.py#L160)
+   for a good example).
+1. The code should be fully understandable, even by a non-native English speaker. This means you should pick
```
Comment (gante) [inline docs/source/add_new_model.mdx:107]:
```suggestion
2. The code should be fully understandable, even by a non-native English speaker. This means you should pick
```

Comment (gante) [reply docs/source/add_new_model.mdx:107]: (looking at the lines below, I think this numbered list needs updated numbers 🙃  )

File: docs/source/add_new_model.mdx [hunk 2]
```diff
@@ -95,6 +95,24 @@ different formats - the model to a *pytorch_model.bin* file and the configuratio
 [`~PretrainedConfig.save_pretrained`], so that both model and configuration are saved.
 
 
+### Code style
+
+When coding your new model, keep in mind that Transformers is an opinionated library and we have a few quirks of our
+own regarding how code should be writte :-)
```
Comment (patil-suraj) [inline docs/source/add_new_model.mdx:101]:
```suggestion
When coding your new model, keep in mind that Transformers is an opinionated library and we have a few quirks of our
own regarding how code should be written :-)
```

File: docs/source/add_new_model.mdx [hunk 3]
```diff
@@ -95,6 +95,24 @@ different formats - the model to a *pytorch_model.bin* file and the configuratio
...
+
+1. The forward pass of your model should be fully written in the modeling file while being fully independent of other
+   models in the library. If you want to reuse a block from another model, copy the code and paste it with a
+   `# Copied from` comment on top (see [here](https://github.com/huggingface/transformers/blob/v4.17.0/src/transformers/models/roberta/modeling_roberta.py#L160)
+   for a good example).
+1. The code should be fully understandable, even by a non-native English speaker. This means you should pick
+   descriptive variable names and avoid abbreviations. As an example, `activation` is preferred to `act`
```
Comment (patil-suraj) [inline docs/source/add_new_model.mdx:108]:
```suggestion
   descriptive variable names and avoid abbreviations. As an example, `activation` is preferred to `act`.
```

File: docs/source/add_new_model.mdx [hunk 4]
```diff
@@ -95,6 +95,24 @@ different formats - the model to a *pytorch_model.bin* file and the configuratio
...
+   One-letter variable names are strongly discouraged unless it's an index in a for loop. 
+1. More generally we prefer longer explicit code to short magical one.
+1. Avoid subclassing `nn.Sequential` in PyTorch but subclass `nn.Module` and write the forward pass, so that anyone
+   using your code can quickly debug it by adding print statement or breaking points
```
Comment (patil-suraj) [inline docs/source/add_new_model.mdx:112]:
```suggestion
   using your code can quickly debug it by adding print statements or breaking points.
```

File: docs/source/add_new_model.mdx [hunk 5]
```diff
@@ -95,6 +95,24 @@ different formats - the model to a *pytorch_model.bin* file and the configuratio
...
+1. Your function signature should be type-annotated. For the rest, good variable names is way more readable and
+   understandable than type annotations.
```
Comment (patil-suraj) [inline docs/source/add_new_model.mdx:114]:
```suggestion
1. Your function signature should be type-annotated. For the rest, good variable names are way more readable and
   understandable than type annotations.
```

**Review by patil-suraj — APPROVED** [review]
Thanks a lot for drafting this!

File: docs/source/add_new_model.mdx [hunk 6]
```diff
@@ -95,6 +95,24 @@ different formats - the model to a *pytorch_model.bin* file and the configuratio
...
+own regarding how code should be written :-)
+
+1. The forward pass of your model should be fully written in the modeling file while being fully independent of other
+   models in the library. If you want to reuse a block from another model, copy the code and paste it with a
+   `# Copied from` comment on top (see [here](https://github.com/huggingface/transformers/blob/v4.17.0/src/transformers/models/roberta/modeling_roberta.py#L160)
+   for a good example).
+2. The code should be fully understandable, even by a non-native English speaker. This means you should pick
+   descriptive variable names and avoid abbreviations. As an example, `activation` is preferred to `act`.
+   One-letter variable names are strongly discouraged unless it's an index in a for loop. 
+1. More generally we prefer longer explicit code to short magical one.
```
Comment (patrickvonplaten) [inline docs/source/add_new_model.mdx:110]:
```suggestion
3. More generally we prefer longer explicit code to short magical one.
```

**Review by patrickvonplaten — APPROVED** [review]

**Review by LysandreJik — APPROVED** [review]
Thanks!

**Pull request merged**